MLFlow Model on MinIO Not Loading

Setup:

- SeldonCore on Kubernetes
- MLFlow (1.10.0) running with MinIO Storage
- Jupyter notebooks on Kubernetes

Jupyter notebook generates a simple sklearn model, it's sent to MLFlow which stores it in MinIO. I'm now trying to get Seldon to create a deployment from this:

```
apiVersion: machinelearning.seldon.io/v1
kind: SeldonDeployment
metadata:
  name: minio-mlflow
spec:
  annotations:
    seldon.io/executor: "true"
  name: newsgroup_nb
  predictors:
  - componentSpecs:
     - spec:
        containers:
         - name: classifier
           livenessProbe:
              failureThreshold: 3
              initialDelaySeconds: 180
              periodSeconds: 5
              successThreshold: 1
              tcpSocket:
                port: http
              timeoutSeconds: 1
    graph:
      children: []
      implementation: MLFLOW_SERVER
      modelUri: s3://mlflow/artifacts/2/adaaee4b5c694f02b5ff9745c53ae75e/artifacts/nb
      envSecretRefName: seldon-init-container-secret
      name: classifier
    name: default
    replicas: 1
```

Container loads up, all the way to:
```[2020-07-29 20:00:51 +0000] [6] [INFO] Listening at: http://0.0.0.0:9000 (6)
[2020-07-29 20:00:51 +0000] [6] [INFO] Using worker: threads
[2020-07-29 20:00:51 +0000] [931] [INFO] Booting worker with pid: 931
2020-07-29 20:00:51,947 - root:load:27 - INFO:  Downloading model from /mnt/models
2020-07-29 20:00:51,947 - root:download:47 - INFO:  Copying contents of /mnt/models to local
[2020-07-29 20:00:51 +0000] [931] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/opt/conda/envs/mlflow/lib/python3.7/site-packages/gunicorn/arbiter.py", line 583, in spawn_worker
    worker.init_process()
  File "/opt/conda/envs/mlflow/lib/python3.7/site-packages/gunicorn/workers/gthread.py", line 92, in init_process
    super().init_process()
  File "/opt/conda/envs/mlflow/lib/python3.7/site-packages/gunicorn/workers/base.py", line 119, in init_process
    self.load_wsgi()
  File "/opt/conda/envs/mlflow/lib/python3.7/site-packages/gunicorn/workers/base.py", line 144, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/opt/conda/envs/mlflow/lib/python3.7/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/opt/conda/envs/mlflow/lib/python3.7/site-packages/seldon_core/app.py", line 71, in load
    self.user_object.load()
  File "/microservice/MLFlowServer.py", line 29, in load
    self._model = pyfunc.load_model(model_folder)
  File "/opt/conda/envs/mlflow/lib/python3.7/site-packages/mlflow/pyfunc/__init__.py", line 297, in load_model
    return importlib.import_module(conf[MAIN])._load_pyfunc(data_path)
  File "/opt/conda/envs/mlflow/lib/python3.7/site-packages/mlflow/sklearn.py", line 230, in _load_pyfunc
    return _load_model_from_local_file(path)
  File "/opt/conda/envs/mlflow/lib/python3.7/site-packages/mlflow/sklearn.py", line 217, in _load_model_from_local_file
    with open(path, "rb") as f:
IsADirectoryError: [Errno 21] Is a directory: '/mnt/models'
[2020-07-29 20:00:51 +0000] [931] [INFO] Worker exiting (pid: 931)
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/opt/conda/envs/mlflow/lib/python3.7/multiprocessing/util.py", line 322, in _exit_function
    p.join()
  File "/opt/conda/envs/mlflow/lib/python3.7/multiprocessing/process.py", line 138, in join
    assert self._parent_pid == os.getpid(), 'can only join a child process'
AssertionError: can only join a child process
[2020-07-29 20:00:51 +0000] [924] [INFO] Handling signal: term
[2020-07-29 20:00:51 +0000] [930] [INFO] Worker exiting (pid: 930)
[2020-07-29 20:00:52 +0000] [924] [INFO] Shutting down: Master
[2020-07-29 20:00:52 +0000] [6] [INFO] Shutting down: Master
[2020-07-29 20:00:52 +0000] [6] [INFO] Reason: Worker failed to boot.
```

Looking at what /mnt/models looks like:
```
bash-4.4$ ls
MLFlowServer.py  before-run  conda_env_create.py  image_metadata.json  license.txt  python  requirements.txt
bash-4.4$ ls /mnt/models/
MLmodel  conda.yaml  model.pkl
```

I can load this successfully via `sk_model = mlflow.sklearn.load_model("s3://mlflow/artifacts/2/adaaee4b5c694f02b5ff9745c53ae75e/artifacts/nb")`

Just not from Seldon. Any ideas? Am I missing something in setting this up?

--- a/.gitignore
+++ b/.gitignore
@@ -100,6 +100,7 @@ examples/models/r_mnist/t10k-labels-idx1-ubyte
 examples/models/r_mnist/train-images-idx3-ubyte
 examples/models/r_mnist/train-labels-idx1-ubyte
 examples/models/sklearn_iris_fbs/IrisClassifier.sav
+examples/models/mlflow_server_ab_test_ambassador/mlruns
 
 # Pickle files
 *.pickle

--- a/examples/models/mlflow_server_ab_test_ambassador/mlflow_server_ab_test_ambassador.ipynb
+++ b/examples/models/mlflow_server_ab_test_ambassador/mlflow_server_ab_test_ambassador.ipynb
@@ -374,8 +374,8 @@
    "outputs": [],
    "source": [
     "!curl -X POST -H 'Content-Type: application/json' \\\n",
-    "    -d \"{'data': {'names': [], 'ndarray': [[7.0, 0.27, 0.36, 20.7, 0.045, 45.0, 170.0, 1.001, 3.0, 0.45, 8.8]]}}\" \\\n",
-    "    http://localhost:80/seldon/seldon/mlflow-deployment/api/v0.1/predictions"
+    "    -d '{\"data\": {\"names\": [], \"ndarray\": [[7.0, 0.27, 0.36, 20.7, 0.045, 45.0, 170.0, 1.001, 3.0, 0.45, 8.8]]}}' \\\n",
+    "    http://localhost:8003/seldon/seldon/mlflow-deployment/api/v0.1/predictions"
    ]
   },
   {
@@ -550,6 +550,12 @@
    "metadata": {},
    "outputs": [],
    "source": [
+    "sc = SeldonClient(\n",
+    "    gateway=\"ambassador\", \n",
+    "    namespace=\"seldon\",\n",
+    "    deployment_name='wines-classifier'\n",
+    ")\n",
+    "\n",
     "def _get_reward(y, y_pred):\n",
     "    if y == y_pred:\n",
     "        return 500    \n",
@@ -568,7 +574,7 @@
     "        data=X,\n",
     "        names=feature_names)\n",
     "    \n",
-    "    y_pred = r.response.data.tensor.values\n",
+    "    y_pred = r.response['data']['tensor']['values']\n",
     "    reward = _get_reward(y, y_pred)\n",
     "    sc.feedback(\n",
     "        deployment_name=\"mlflow-deployment\",\n",
@@ -625,7 +631,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.6.8"
+   "version": "3.7.8"
   },
   "varInspector": {
    "cols": {

--- a/servers/mlflowserver/mlflowserver/MLFlowServer.py
+++ b/servers/mlflowserver/mlflowserver/MLFlowServer.py
@@ -1,13 +1,14 @@
-import numpy as np
+import yaml
+import os
 import logging
 import requests
+import numpy as np
 import pandas as pd
+
 from mlflow import pyfunc
 from seldon_core import Storage
 from seldon_core.user_model import SeldonComponent
-from typing import Dict, List, Union, Iterable
-import yaml
-import os
+from typing import Dict, List, Union
 
 logger = logging.getLogger()
 
@@ -15,7 +16,7 @@ MLFLOW_SERVER = "model"
 
 
 class MLFlowServer(SeldonComponent):
-    def __init__(self, model_uri: str, xtype: str = 'ndarray'):
+    def __init__(self, model_uri: str, xtype: str = "ndarray"):
         super().__init__()
         logger.info(f"Creating MLFLow server with URI {model_uri}")
         logger.info(f"xtype: {xtype}")
@@ -30,9 +31,9 @@ class MLFlowServer(SeldonComponent):
         self.ready = True
 
     def predict(
-        self, X: np.ndarray, feature_names: Iterable[str] = [], meta: Dict = None
+        self, X: np.ndarray, feature_names: List[str] = [], meta: Dict = None
     ) -> Union[np.ndarray, List, Dict, str, bytes]:
-        logger.info(f"Requesting prediction with: {X}")
+        logger.debug(f"Requesting prediction with: {X}")
 
         if not self.ready:
             raise requests.HTTPError("Model not loaded yet")
@@ -45,7 +46,8 @@ class MLFlowServer(SeldonComponent):
             else:
                 df = pd.DataFrame(data=X)
             result = self._model.predict(df)
-        logger.info(f"Prediction result: {result}")
+
+        logger.debug(f"Prediction result: {result}")
         return result
 
     def init_metadata(self):

--- a/servers/mlflowserver/mlflowserver/requirements.txt
+++ b/servers/mlflowserver/mlflowserver/requirements.txt
@@ -1,7 +1,7 @@
-pyyaml==5.1.2
-requests==2.22.0
-mlflow>=1.10.0,<1.12.0
-pandas
+pyyaml<5.4.0
+requests<2.25.0
+mlflow<1.12.0
+pandas<1.2.0
 
 # local seldon-core inside the s2i image
 /microservice/python

--- a/servers/mlflowserver/mlflowserver/requirements.txt
+++ b/servers/mlflowserver/mlflowserver/requirements.txt
@@ -1,6 +1,6 @@
 pyyaml==5.1.2
 requests==2.22.0
-mlflow>=1.4.0,<1.9.0
+mlflow>=1.10.0,<1.12.0
 pandas
 
 # local seldon-core inside the s2i image

